File: Treino/Semana 1/Daniel/fliper_v2.py
constants_nodes = (
    'A',
    'B',
    'C',
    'R',
    'P'
)

numbers_nodes = len(constants_nodes)
adj = {}
parent = {}
for i in range(numbers_nodes):
    parent[constants_nodes[i]] = -1

input_line = input().split(' ')
is_right_r = input_line[1] == '1'
is_right_p = input_line[0] == '1'

def dfs(node) -> None:
    if adj[node]:
        for next_node in adj[node]:
            if parent[next_node] != -1:
                continue
            parent[next_node] = node
            dfs(next_node)
    else:
        print(node)

for i in range(numbers_nodes):
    adj[constants_nodes[i]] = []

if is_right_p:
    adj['P'].append('R')
else:
    adj['P'].append('C')

if is_right_r:
    adj['R'].append('A')
else:
    adj['R'].append('B')

dfs('P')

File: Treino/Semana 1/fliper_v2.cpp
#include <iostream>
#include <vector>

#define MAX_NODES 1000
#define nodeP 0
#define nodeR 1
#define nodeA 2
#define nodeB 3
#define nodeC 4

using namespace std;

vector <int> adj[MAX_NODES];
int parent[MAX_NODES];
bool isRightP, isRightR;
int answer;

void dfs(int node)
{
    if (adj[node].empty())
    {
        answer = node;
    }
    else
    {
        for (auto nextNode: adj[node])
        {
            if (parent[nextNode] != -1) continue;
            parent[nextNode] = node;
            dfs(nextNode);
        }
    }
}

int main()
{
    for (int i = 0; i < MAX_NODES; i++) parent[i] = -1;
    cin >> isRightP >> isRightR;

    if (isRightP)
    {
        adj[nodeP].push_back(nodeR);
    }
    else
    {
        adj[nodeP].push_back(nodeC);
    }

    if (isRightR)
    {
        adj[nodeR].push_back(nodeA);
    }
    else
    {
        adj[nodeR].push_back(nodeB);
    }

    dfs(nodeP);

    switch (answer)
    {
        case nodeA:
            cout << "A";
            break;
        
        case nodeB:
            cout << "B";
            break;

        case nodeC:
            cout << "C";
            break;

        default:
            cout << "Error";
            break;
    }
    return 0;
}